### PR TITLE
Track per size class internal fragmentation

### PR DESCRIPTION
**Pull request by yinan1048576**
Add an opt-in option to track internal fragmentation for prof-sampled allocations. The internal fragmentation is represented by a tuple of `(total_request_size, count)`. The caller can compute the exact internal fragmentation using the formula `1 - (double)(total_request_size) / (count * usize)`.

The tuple can be obtained either via `mallctl` or in the stats print. The raw tuple instead of the actual internal fragmentation number is reported to minimize floating-point complications to the `mallctl` and stats modules. The `count` in the tuple is also a nice number to know itself, for e.g. verifying the actual proportion of sampled allocations.

The internal fragmentation is reported in a per size class manner. It's possible to have finer grained reports: the recording table can just have more rows, but the format of the `mallctl` and the stats print can be slightly trickier, so don't bother doing it for now.

### CODE REVIEW

**Comment by yinan1048576** [comment]
Revised the stats output. The tricky part is: for sampled small allocations, the prof stats are recorded for the small sizes whereas the other stats are recorded for the smallest large size class, to which the small size is bumped. Had to have a special way to handle it.

**Comment by yinan1048576** [comment]
Added more tests around aligned allocations, and rebase after the #1926 fix.

**Comment by yinan1048576** [comment]
Fix test failures.

**Comment by yinan1048576** [comment]
Expand the stats field width for the accumulated request size, as it was observed to go beyond the current room.

File: src/stats.c [hunk 1]
```diff
@@ -556,10 +635,34 @@ stats_arena_lextents_print(emitter_t *emitter, unsigned i, uint64_t uptime) {
...
+			xmallctlbymib(prof_live_mib, prof_stats_miblen,
+			    &prof_live, &prof_stats_len, NULL, 0);
+			prof_accum_mib[3] = j;
+			xmallctlbymib(prof_accum_mib, prof_stats_miblen,
+			    &prof_accum, &prof_stats_len, NULL, 0);
+		}
+
 		emitter_json_object_begin(emitter);
+		if (prof_stats_on) {
+			emitter_json_kv(emitter, "prof_live_r",
```
Comment (davidtgoldblatt) [inline src/stats.c:657]: Probably "prof_live_requested" and "prof_live_count" are better names for these keys?

File: test/unit/prof_stats.c [hunk 2]
```diff
@@ -76,28 +81,90 @@ test_wrapper(szind_t ind) {
...
+	test_szind_wrapper(SC_NBINS);
+	test_szind_wrapper(SC_NBINS + 1);
+	test_szind_wrapper(SC_NBINS + 2);
+}
+TEST_END
+
+static void
+test_szind_bins_aligned_wrapper(szind_t ind) {
+	unsigned alignment = (unsigned)sz_index2size(ind);
+	if (ffs_u(alignment) != fls_u(alignment)) {
```
[the remark this answers is not in the document]
Comment (yinan1048576) [reply test/unit/prof_stats.c:117]: Alignment does change things in the test: the sizes are 1, 2, and 3, and the alignments are 8, 16 and 32. But manually specifying the alignment is a good idea.

File: src/stats.c (hunk 1 above)
Comment (yinan1048576) [reply src/stats.c:657]: Yes. I was trying to make the column narrower, but, you're right, the names became too hard to understand. Expanded the name, as well as the column widths.

**Review by yinan1048576 — COMMENTED** [review]
Rebase, and revise according to feedback. Also a couple of other minor improvements along the way.

File: include/jemalloc/internal/prof_stats.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#ifndef JEMALLOC_INTERNAL_PROF_STATS_H
+#define JEMALLOC_INTERNAL_PROF_STATS_H
+
+typedef struct prof_stats_s prof_stats_t;
+struct prof_stats_s {
+	uint64_t req_sum;
+	uint64_t count;
+};
+
+extern malloc_mutex_t prof_stats_mtx;
```
[the remark this answers is not in the document]
Comment (davidtgoldblatt) [reply include/jemalloc/internal/prof_stats.h:10]: Yeah, but given how easy it is to do (see e.g. https://github.com/jemalloc/jemalloc/commit/484f04733e5bd9908faf502fced6df66ca33f9f9), we should default to always including the stats for new mutexes.

File: src/jemalloc.c [hunk 4]
```diff
@@ -1502,6 +1502,7 @@ malloc_conf_init_helper(sc_data_t *sc_data, unsigned bin_shard_sizes[SC_NBINS],
 				CONF_HANDLE_BOOL(opt_prof_log, "prof_log")
 				CONF_HANDLE_SSIZE_T(opt_prof_recent_alloc_max,
 				    "prof_recent_alloc_max", -1, SSIZE_MAX)
+				CONF_HANDLE_BOOL(opt_prof_stats, "prof_stats")
```
[the remark this answers is not in the document]
Comment (davidtgoldblatt) [reply src/jemalloc.c:1555]: Sorry, was thinking stats.

File: src/ctl.c [hunk 5]
```diff
@@ -131,6 +131,7 @@ CTL_PROTO(opt_prof_final)
 CTL_PROTO(opt_prof_leak)
 CTL_PROTO(opt_prof_accum)
 CTL_PROTO(opt_prof_recent_alloc_max)
+CTL_PROTO(opt_prof_stats)
```
[the remark this answers is not in the document]
Comment (davidtgoldblatt) [reply src/ctl.c:140]: Yeah; this is a man-page guarantee.

(I hate hate hate how much boilerplate there is to keep all of stats / conf / ctl in sync; I've got some refactoring ideas here though).

**Comment by yinan1048576** [comment]
Added a separate commit to include the mutex stats.

**Review by davidtgoldblatt — APPROVED** [review]

**Comment by yinan1048576** [comment]
Rebase.

**Pull request merged**